# PES University, Bangalore
Established under Karnataka Act No. 16 of 2013

## UE22AM343AB4 - Advanced Data Analytics

Designed by Sathwik HJ

### Student Details
- Name : **Shusrith S**
- SRN : **PES1UG22AM155**

## Market Basket Analysis
    + Apriori Algorithm
    + Association Rule mining

### Prerequisites
- Revise the following concepts
    - Apriori Algorithm
        - Suport
        - Confidence
        - Lift
- Install the following software
    - pandas
    - apyori

### Problems

1. Problem 1: Preprocessing
2. Problem 2: Item set detection and analaysis
3. Problem 3: Association rule minning
- Note: Make sure to include justifications for each of your approaches and also provide a brief observations of the results obtained

### Context
Welcome to "Shocking Electronics", a budding Electronic Store. As a data analyst working with "Shocking Electronics" , your mission is to uncover meaningful patterns within customer transactions to enhance their shopping experience and help us compete with our competitor "Circuit Circus Electronics".

In [1]:
!wget https://raw.githubusercontent.com/MBUYt0n/ada/main/ADA_Worksheet_C/Electronics.csv

--2024-09-20 16:25:37--  https://raw.githubusercontent.com/MBUYt0n/ada/main/ADA_Worksheet_C/Electronics.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.110.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 9593 (9.4K) [text/plain]
Saving to: ‘Electronics.csv’

Electronics.csv     100%[===================>]   9.37K  --.-KB/s    in 0s      

2024-09-20 16:25:37 (103 MB/s) - ‘Electronics.csv’ saved [9593/9593]



### About the dataset
- "Electronics.csv"
- Each record of the dataset represents a transaction made by a customer at "Shocking Electronics store".
- The transaction contains the items bought in that transaction

In [2]:
# Install apyori package
# Use % or ! based on the environment to install
%pip install apyori
%pip install mlxtend
import apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5953 sha256=49fa3adaf93915eee1e6325eb5145c709005194a547e9ac2fc43b8323652480a
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


apyori documentation: https://pypi.org/project/apyori/ (Refer API Usage)

In [3]:
import pandas as pd
from apyori import apriori

### Problem 1 - Preprocessing
**Load the transactions data** from the provided csv file. **Transform the data** to a suitable format (Hint: **List of lists**[internal list contains the items of the transaction]). **Make sure to clean the data** (Hint: NA values).


In [4]:
df = pd.read_csv('Electronics.csv', names=["item1", "item2", "item3", "item4", "item5", "item6"], index_col=False)
df

item1                item2                item3  \
0          smartphone              earbuds     screen protector   
1              laptop                mouse             keyboard   
2          smartwatch      fitness tracker     wireless charger   
3                  TV             soundbar           HDMI cable   
4              tablet               stylus          tablet case   
..                ...                  ...                  ...   
212  desktop computer  mechanical keyboard         gaming mouse   
213        smartwatch      fitness tracker        charging dock   
214            laptop               webcam  external hard drive   
215            tablet     screen protector               stylus   
216  desktop computer       gaming monitor         RGB keyboard   

                   item4                item5       item6  
0             power bank            USB cable  phone case  
1           laptop stand  external hard drive         NaN  
2                    NaN                  NaN         NaN  
3       streaming device                  NaN         NaN  
4         screen cleaner                  NaN         NaN  
..                   ...                  ...         ...  
212                  NaN                  NaN         NaN  
213  wireless headphones                  NaN         NaN  
214              USB hub                  NaN         NaN  
215     portable charger                  NaN         NaN  
216               webcam                  NaN         NaN  

[217 rows x 6 columns]

In [5]:
df.isna().sum()

item1      0
item2      0
item3      5
item4     86
item5    213
item6    215
dtype: int64

In [6]:
df.fillna("", axis=1, inplace=True)
df

item1                item2                item3  \
0          smartphone              earbuds     screen protector   
1              laptop                mouse             keyboard   
2          smartwatch      fitness tracker     wireless charger   
3                  TV             soundbar           HDMI cable   
4              tablet               stylus          tablet case   
..                ...                  ...                  ...   
212  desktop computer  mechanical keyboard         gaming mouse   
213        smartwatch      fitness tracker        charging dock   
214            laptop               webcam  external hard drive   
215            tablet     screen protector               stylus   
216  desktop computer       gaming monitor         RGB keyboard   

                   item4                item5       item6  
0             power bank            USB cable  phone case  
1           laptop stand  external hard drive              
2                                                          
3       streaming device                                   
4         screen cleaner                                   
..                   ...                  ...         ...  
212                                                        
213  wireless headphones                                   
214              USB hub                                   
215     portable charger                                   
216               webcam                                   

[217 rows x 6 columns]

In [7]:
import numpy as np
len(set(np.ravel(df.values)))

93

In [8]:
items = df.apply(lambda x: [i for i in x if i != ''], axis=1).tolist()
items

[['smartphone',
  'earbuds',
  'screen protector',
  'power bank',
  'USB cable',
  'phone case'],
 ['laptop', 'mouse', 'keyboard', 'laptop stand', 'external hard drive'],
 ['smartwatch', 'fitness tracker', 'wireless charger'],
 ['TV', 'soundbar', 'HDMI cable', 'streaming device'],
 ['tablet', 'stylus', 'tablet case', 'screen cleaner'],
 ['smartphone', 'smartwatch', 'earbuds'],
 ['gaming console', 'game controller', 'HDMI cable', 'gaming headset'],
 ['laptop', 'wireless mouse', 'backpack', 'laptop sleeve'],
 ['smartphone', 'phone case', 'screen protector', 'USB cable'],
 ['smartphone', 'smartwatch', 'power bank'],
 ['laptop', 'external hard drive', 'USB hub'],
 ['TV', 'soundbar', 'HDMI cable'],
 ['smartphone', 'earbuds', 'power bank'],
 ['smartwatch', 'fitness tracker', 'charging dock'],
 ['gaming console', 'game controller', 'gaming headset'],
 ['laptop', 'keyboard', 'mouse', 'monitor'],
 ['tablet', 'stylus', 'tablet case'],
 ['smartphone', 'screen protector', 'phone case'],
 ['TV', '

### Problem 2 - Item sets
1. Print out the frequent item sets along with their support values also display the count of item sets.(**min_support=0.08**)
2. "Shocking Electronics" wants to replenish its stocks, help find the top 5 most popular(higher buying frequency) items/item_sets to replenish. **Explain and justify** the process followed to come to the conclusion.


In [9]:
from apyori import apriori

frequent_itemsets = apriori(items, min_support=0.08)

frequent_itemsets_list = list(frequent_itemsets)
for itemset in frequent_itemsets_list:
    print(f"Itemset: {itemset.items}, Support: {itemset.support}")


Itemset: frozenset({'HDMI cable'}), Support: 0.15668202764976957
Itemset: frozenset({'TV'}), Support: 0.1382488479262673
Itemset: frozenset({'earbuds'}), Support: 0.1152073732718894
Itemset: frozenset({'external hard drive'}), Support: 0.1382488479262673
Itemset: frozenset({'fitness tracker'}), Support: 0.11981566820276497
Itemset: frozenset({'gaming console'}), Support: 0.12903225806451613
Itemset: frozenset({'gaming headset'}), Support: 0.08755760368663594
Itemset: frozenset({'laptop'}), Support: 0.1889400921658986
Itemset: frozenset({'phone case'}), Support: 0.09216589861751152
Itemset: frozenset({'screen protector'}), Support: 0.15207373271889402
Itemset: frozenset({'smartphone'}), Support: 0.2350230414746544
Itemset: frozenset({'smartwatch'}), Support: 0.14285714285714285
Itemset: frozenset({'soundbar'}), Support: 0.12442396313364056
Itemset: frozenset({'streaming device'}), Support: 0.0967741935483871
Itemset: frozenset({'stylus'}), Support: 0.09216589861751152
Itemset: frozenset

In [10]:
item_support = {}
for itemset in frequent_itemsets_list:
    for item in itemset.items:
        item_support[item] = item_support.get(item, 0) + itemset.support

top_items = sorted(item_support.items(), key=lambda x: x[1], reverse=True)[:5]

print("Top 5 items to replenish:")
for item, support in top_items:
    print(f"Item: {item}, Support: {support}")


Top 5 items to replenish:
Item: smartphone, Support: 0.5483870967741935
Item: TV, Support: 0.45161290322580644
Item: soundbar, Support: 0.43317972350230416
Item: laptop, Support: 0.4147465437788018
Item: streaming device, Support: 0.3778801843317972


These are the top 5 items to replenish.

The process to identify popular itemsets for replenishment is as follows:

1. **Frequent Itemset Mining:** The Apriori algorithm, with a minimum support threshold of 0.08, identified itemsets appearing in at least 8% of transactions.
2. **Support Calculation:** Support values, indicating the percentage of transactions containing each itemset, were computed.
3. **Sorting by Support:** Itemsets were sorted by support in descending order to prioritize popularity.
4. **Top 5 Selection:** The top 5 itemsets with the highest support were chosen for replenishment.

This approach ensures that "Shocking Electronics" stocks the most popular products and combinations, maximizing customer satisfaction and sales.

### Problem 3 - Association Rules
+ Items on the left side of the association rule are called : **Antecedent items** and the right side of the association rule are called : **Consequent Items**.
1. Print out the association rules along with their confidence and lift. (Analyse the output structure of apriori())
    + **(min_support=0.01, min_confidence = 0.08, min_lift=1.7, min_length=3)**
2. As the festive season is approaching, "Shocking Electronics" is considering to provide discounts and offers on some of their products. Help them identify the top 5 popular **pairs/sets** of items/item_sets bought, considering probability of consequent item being purchased when antecedent item is bought.
3. Also help them identify the top 5 popular **pairs/sets** of items/item_sets bought together, considering the popularity of consequent and antecedent items.
+ (Consequent and antecedent items together form the **pairs/sets** specified in the question)


In [11]:
# Let's get association rules along with confidence and lift
association_rules = []
for result in frequent_itemsets_list:
    for ordered_statistic in result.ordered_statistics:
        antecedent = list(ordered_statistic.items_base)
        consequent = list(ordered_statistic.items_add)
        support = result.support
        confidence = ordered_statistic.confidence
        lift = ordered_statistic.lift

        association_rules.append((antecedent, consequent, support, confidence, lift))

rules_df = pd.DataFrame(association_rules, columns=['Antecedent', 'Consequent', 'Support', 'Confidence', 'Lift'])
rules_df

Antecedent                        Consequent   Support  \
0                             []                      [HDMI cable]  0.156682   
1                             []                              [TV]  0.138249   
2                             []                         [earbuds]  0.115207   
3                             []             [external hard drive]  0.138249   
4                             []                 [fitness tracker]  0.119816   
5                             []                  [gaming console]  0.129032   
6                             []                  [gaming headset]  0.087558   
7                             []                          [laptop]  0.188940   
8                             []                      [phone case]  0.092166   
9                             []                [screen protector]  0.152074   
10                            []                      [smartphone]  0.235023   
11                            []                      [smartwatch]  0.142857   
12                            []                        [soundbar]  0.124424   
13                            []                [streaming device]  0.096774   
14                            []                          [stylus]  0.092166   
15                            []                          [tablet]  0.087558   
16                            []                  [wireless mouse]  0.092166   
17                            []                    [TV, soundbar]  0.124424   
18                          [TV]                        [soundbar]  0.124424   
19                    [soundbar]                              [TV]  0.124424   
20                            []            [TV, streaming device]  0.096774   
21                          [TV]                [streaming device]  0.096774   
22            [streaming device]                              [TV]  0.096774   
23                            []             [smartphone, earbuds]  0.101382   
24                     [earbuds]                      [smartphone]  0.101382   
25                  [smartphone]                         [earbuds]  0.101382   
26                            []     [laptop, external hard drive]  0.133641   
27         [external hard drive]                          [laptop]  0.133641   
28                      [laptop]             [external hard drive]  0.133641   
29                            []     [fitness tracker, smartwatch]  0.115207   
30             [fitness tracker]                      [smartwatch]  0.115207   
31                  [smartwatch]                 [fitness tracker]  0.115207   
32                            []  [gaming console, gaming headset]  0.082949   
33              [gaming console]                  [gaming headset]  0.082949   
34              [gaming headset]                  [gaming console]  0.082949   
35                            []          [laptop, wireless mouse]  0.092166   
36                      [laptop]                  [wireless mouse]  0.092166   
37              [wireless mouse]                          [laptop]  0.092166   
38                            []          [smartphone, phone case]  0.092166   
39                  [phone case]                      [smartphone]  0.092166   
40                  [smartphone]                      [phone case]  0.092166   
41                            []    [smartphone, screen protector]  0.119816   
42            [screen protector]                      [smartphone]  0.119816   
43                  [smartphone]                [screen protector]  0.119816   
44                            []      [soundbar, streaming device]  0.092166   
45                    [soundbar]                [streaming device]  0.092166   
46            [streaming device]                        [soundbar]  0.092166   
47                            []                  [tablet, stylus]  0.082949   
48                      [stylus]                          [tablet]  0.082949   
49                    

In [12]:
min_confidence = 0.08
min_lift = 1.7
filtered_rules_df = rules_df[(rules_df['Confidence'] >= min_confidence) & (rules_df['Lift'] >= min_lift)]
filtered_rules_df

Antecedent                    Consequent   Support  \
18                          [TV]                    [soundbar]  0.124424   
19                    [soundbar]                          [TV]  0.124424   
21                          [TV]            [streaming device]  0.096774   
22            [streaming device]                          [TV]  0.096774   
24                     [earbuds]                  [smartphone]  0.101382   
25                  [smartphone]                     [earbuds]  0.101382   
27         [external hard drive]                      [laptop]  0.133641   
28                      [laptop]         [external hard drive]  0.133641   
30             [fitness tracker]                  [smartwatch]  0.115207   
31                  [smartwatch]             [fitness tracker]  0.115207   
33              [gaming console]              [gaming headset]  0.082949   
34              [gaming headset]              [gaming console]  0.082949   
36                      [laptop]              [wireless mouse]  0.092166   
37              [wireless mouse]                      [laptop]  0.092166   
39                  [phone case]                  [smartphone]  0.092166   
40                  [smartphone]                  [phone case]  0.092166   
42            [screen protector]                  [smartphone]  0.119816   
43                  [smartphone]            [screen protector]  0.119816   
45                    [soundbar]            [streaming device]  0.092166   
46            [streaming device]                    [soundbar]  0.092166   
48                      [stylus]                      [tablet]  0.082949   
49                      [tablet]                      [stylus]  0.082949   
51                          [TV]  [soundbar, streaming device]  0.092166   
52                    [soundbar]        [TV, streaming device]  0.092166   
53            [streaming device]                [TV, soundbar]  0.092166   
54                [TV, soundbar]            [streaming device]  0.092166   
55        [TV, streaming device]                    [soundbar]  0.092166   
56  [streaming device, soundbar]                          [TV]  0.092166   

    Confidence       Lift  
18    0.900000   7.233333  
19    1.000000   7.233333  
21    0.700000   7.233333  
22    1.000000   7.233333  
24    0.880000   3.744314  
25    0.431373   3.744314  
27    0.966667   5.116260  
28    0.707317   5.116260  
30    0.961538   6.730769  
31    0.806452   6.730769  
33    0.642857   7.342105  
34    0.947368   7.342105  
36    0.487805   5.292683  
37    1.000000   5.292683  
39    1.000000   4.254902  
40    0.392157   4.254902  
42    0.787879   3.352347  
43    0.509804   3.352347  
45    0.740741   7.654321  
46    0.952381   7.654321  
48    0.900000  10.278947  
49    0.947368  10.278947  
51    0.666667   7.233333  
52    0.740741   7.654321  
53    0.952381   7.654321  
54    0.740741   7.654321  
55    0.952381   7.654321  
56    1.000000   7.233333

In [13]:
rules = []
for itemset in frequent_itemsets_list:
    for ordered_stat in itemset.ordered_statistics:
        if len(ordered_stat.items_base) > 0 and len(ordered_stat.items_add) > 0:
            rules.append({
                'antecedent': tuple(ordered_stat.items_base),
                'consequent': tuple(ordered_stat.items_add),
                'support': itemset.support,
                'confidence': ordered_stat.confidence,
                'lift': ordered_stat.lift
            })

rules_sorted_by_confidence = sorted(rules, key=lambda x: x['confidence'], reverse=True)[:5]
rules_sorted_by_lift = sorted(rules, key=lambda x: x['lift'], reverse=True)[:5]

print("Top 5 popular pairs/sets based on confidence:")
for rule in rules_sorted_by_confidence:
    print(f"Antecedent: {rule['antecedent']}, Consequent: {rule['consequent']}, Confidence: {rule['confidence']:.2f}, Lift: {rule['lift']:.2f}")

print("\nTop 5 popular pairs/sets based on lift:")
for rule in rules_sorted_by_lift:
    print(f"Antecedent: {rule['antecedent']}, Consequent: {rule['consequent']}, Confidence: {rule['confidence']:.2f}, Lift: {rule['lift']:.2f}")


Top 5 popular pairs/sets based on confidence:
Antecedent: ('soundbar',), Consequent: ('TV',), Confidence: 1.00, Lift: 7.23
Antecedent: ('streaming device',), Consequent: ('TV',), Confidence: 1.00, Lift: 7.23
Antecedent: ('wireless mouse',), Consequent: ('laptop',), Confidence: 1.00, Lift: 5.29
Antecedent: ('phone case',), Consequent: ('smartphone',), Confidence: 1.00, Lift: 4.25
Antecedent: ('streaming device', 'soundbar'), Consequent: ('TV',), Confidence: 1.00, Lift: 7.23

Top 5 popular pairs/sets based on lift:
Antecedent: ('stylus',), Consequent: ('tablet',), Confidence: 0.90, Lift: 10.28
Antecedent: ('tablet',), Consequent: ('stylus',), Confidence: 0.95, Lift: 10.28
Antecedent: ('streaming device',), Consequent: ('soundbar',), Confidence: 0.95, Lift: 7.65
Antecedent: ('streaming device',), Consequent: ('TV', 'soundbar'), Confidence: 0.95, Lift: 7.65
Antecedent: ('TV', 'streaming device'), Consequent: ('soundbar',), Confidence: 0.95, Lift: 7.65


### Problem 4 - Theory
Explain **Rule**, **Lift** and **Support**, taking examples from the results obtained from **your analysis**



1. **Rule**:
   - An **association rule** is a relationship between two or more items. It helps in discovering how items are associated with one another in a dataset. A rule is structured in the form `Antecedent → Consequent`, where the antecedent is the item(s) purchased first, and the consequent is the item(s) purchased after.
   
   **Example**:
   - **Rule**: `('soundbar',) → ('TV',)`
     - This rule indicates that customers who buy a soundbar also tend to buy a TV.

2. **Support**:
   - **Support** measures the frequency of occurrence of an itemset in the dataset. It is the ratio of transactions that contain the itemset to the total number of transactions. It helps determine how popular an itemset is.
   
   **Formula**:  
   \[
   \text{Support}(X \rightarrow Y) = \frac{\text{Transactions containing } X \text{ and } Y}{\text{Total transactions}}
   \]
   
   **Example**:
   - If the **support** for the rule `('soundbar',) → ('TV',)` is `0.15`, it means that 15% of all transactions include both a soundbar and a TV.

3. **Lift**:
   - **Lift** measures how much more likely the consequent is to be purchased when the antecedent is purchased, compared to the chance of purchasing the consequent independently. It indicates the strength of the association.
   
   **Formula**:  
   \[
   \text{Lift}(X \rightarrow Y) = \frac{\text{Confidence}(X \rightarrow Y)}{\text{Support}(Y)}
   \]
   - A **lift** greater than 1 indicates a strong positive association between the antecedent and the consequent.
   
   **Example**:
   - **Lift** for the rule `('soundbar',) → ('TV',)` is `7.23`, meaning that customers who buy a soundbar are 7.23 times more likely to buy a TV compared to customers who don't buy a soundbar.

